grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -444,42 +449,97 @@ private static String getResponseBody(List<FieldDescriptor> topLevelFields,
-                                fieldName.equals(responseBody)) {
-                                final byte[] bytes = mapper.writeValueAsBytes(responseBodyJsonNode);
-                                return HttpData.wrap(bytes);
-                            }
-                        }
-                        return HttpData.ofUtf8("null");
-                    } catch (IOException e) {
-                        logger.warn("Unexpected exception while extracting responseBody '{}' from {}",
-                                    responseBody, data, e);
-                        return HttpData.wrap(array);
+    private static Function<AggregatedHttpResponse, AggregatedHttpResponse> generateResponseConverter(
+            TranscodingSpec spec) {
+        // Ignore the spec if the method is HttpBody. The response body is already in the correct format.
+        if (HttpBody.getDescriptor().equals(spec.methodDescriptor.getOutputType())) {
+            return httpResponse -> {
+                try (HttpData data = httpResponse.content()) {
+                    final JsonNode jsonNode = extractHttpBody(data);
+
+                    // Failed to parse the JSON body, return the original response.
+                    if (jsonNode == null) {
+                        return httpResponse;
                     }
+
+                    // The data field is base64 encoded.
+                    // https://protobuf.dev/programming-guides/proto3/#json
+                    final String httpBody = jsonNode.get("data").asText();
+                    final byte[] httpBodyBytes = Base64.getDecoder().decode(httpBody);
+
+                    final ResponseHeaders newHeaders = httpResponse.headers().withMutations(builder -> {
+                        final JsonNode contentType = jsonNode.get("contentType");
+
+                        if (contentType != null && contentType.isTextual()) {
+                            builder.set(HttpHeaderNames.CONTENT_TYPE, contentType.textValue());
+                        } else {
+                            builder.remove(HttpHeaderNames.CONTENT_TYPE);
+                        }
+                        builder.contentLength(httpBodyBytes.length);
+                    });
+
+                    return AggregatedHttpResponse.of(newHeaders, HttpData.wrap(httpBodyBytes));
                 }
             };
         }
+
+        @Nullable
+        final String responseBody = spec.responseBody;
+        if (responseBody == null) {
+            return null;
+        }
+
+        return httpResponse -> {
+            try (HttpData data = httpResponse.content()) {
+                final HttpData convertedData = convertHttpDataForResponseBody(responseBody, data);
+                final ResponseHeaders newHeaders = httpResponse.headers().withMutations(builder -> {
+                    builder.contentLength(convertedData.length());
+                });
+                return AggregatedHttpResponse.of(newHeaders, convertedData);
```

Comment:
nit; we can probably simplify to just this since the `AggregatedHttpResponse#of` method will set the content length for us

```suggestion
                return AggregatedHttpResponse.of(httpResponse.headers(), convertedData);
```

Reply:
Done.